### PR TITLE
Read immutable set

### PR DESCRIPTION
**Pull request by original-brownbear**
Just like the readImmutableList method, adding this avoids some copying
and needless wrapping relative to the replace usages.
The few plain usages of the mutable set read that were replaced are
safe to replace since their non-stream-input constructors use immutable
set copies.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 1]
```diff
@@ -1159,6 +1159,32 @@ public <T> Set<T> readSet(Writeable.Reader<T> reader) throws IOException {
...
+        } else if (count == 2) {
+            return Set.of(reader.read(this), reader.read(this));
+        }
+        Object[] entries = new Object[count];
+        for (int i = 0; i < count; i++) {
+            entries[i] = reader.read(this);
+        }
+        @SuppressWarnings("unchecked")
+        T[] typedEntries = (T[]) entries;
+        return Set.of(typedEntries);
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:1185]: `Set.of` throws an exception on duplicates on contrary to `Set.copyOf`. I wonder if we should document it in the method Javadoc to make sure that we can use this method if the written data is a `Set`.

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 2]
```diff
@@ -1159,6 +1159,32 @@ public <T> Set<T> readSet(Writeable.Reader<T> reader) throws IOException {
...
+     * Reads a set of objects. The set is expected to have been written using {@link StreamOutput#writeCollection(Collection)}}.
+     * The returned set is immutable.
+     *
+     * @return the set of objects
+     * @throws IOException if an I/O exception occurs reading the set
+     */
+    public <T> Set<T> readImmutableSet(final Writeable.Reader<T> reader) throws IOException {
+        int count = readArraySize();
+        // special cases small arrays, just like in java.util.Set.of(...)
+        if (count == 0) {
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:1172]: NIT: this could be simplified with a switch expression

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:1172]: Hmm not really simpler here, it's far more LoC with a switch I think :)

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:1185]: Sure thing added a couple words on that.

**Review by arteam — APPROVED** [review]
LGTM. Thanks Armin

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 2 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:1172]: I do not think so:
```
        return switch (count) {
            case 0 -> Set.of();
            case 1 -> Set.of(reader.read(this));
            case 2 -> Set.of(reader.read(this), reader.read(this));
            default -> {
                Object[] entries = new Object[count];
                for (int i = 0; i < count; i++) {
                    entries[i] = reader.read(this);
                }
                @SuppressWarnings("unchecked")
                T[] typedEntries = (T[]) entries;
                yield  Set.of(typedEntries);
            }
        };
```

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 3]
```diff
@@ -1159,6 +1159,32 @@ public <T> Set<T> readSet(Writeable.Reader<T> reader) throws IOException {
         return readCollection(reader, Sets::newHashSetWithExpectedSize, Collections.emptySet());
     }
 
+    /**
+     * Reads a set of objects. The set is expected to have been written using {@link StreamOutput#writeCollection(Collection)}} with
+     * a collection that contains no duplicates. The returned set is immutable.
+     *
+     * @return the set of objects
+     * @throws IOException if an I/O exception occurs reading the set
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:1167]: Lets also document `IllegalArgumentException("duplicate element: " + e);` so it has more chances to be noticed here

**Review by idegtiarenko — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: huh fancy :D still the same number of code lines though, looks nice though so I went with this and used it in the duplicate list code also :)

**Comment by original-brownbear** [comment]
Thanks both!

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1

**Pull request merged**